### PR TITLE
Add a domainShift in ImageContainerByITKImage

### PR DESCRIPTION
**Pull request by phcerdan**
Allowing to shift the domain, without affecting the indexing
of the underlying ITK image.

Also add basic conversions from index points to physical points.

The `domainShift` parameter can only be set with the existing
function`updateDomain`. It defaults to `Zero`.

The implementation provided in #1484 was untested,
and I would say it goes out of the buffer of the
ITK region when using `operator()` or `setValue`.

This patch introduces a conversion function between what I call a
`domainPoint` (a point in the DGtal domain defined by `lowerBound` and `upperBound`),
and an `indexPoint` which is a valid point in the ITK region.

These two are equivalent when `domainShift` is Zero (the default).

It also effectively reverts #1484 that mixes index space and physical
space (with origins) in ITK.

Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [NA] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [ ] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [ ] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

**Comment by kerautret** [comment]
@phcerdan sorry if the patch make some issues for you but for me it was just to be able to shift the image domain in relation physical origin defined in some type ITK. Else the imported image was wrongly always centered in 0, 0, 0. What type of issue did you have ? I use the patch with a lot of image and no pb. Can you give me a link of the image where it fails ? Or perhaps can add an option to let the user centered or not the domain (Ie in updateDomain).

**Comment by phcerdan** [comment]
My use case is simple, I have two images.
The first has metadata on it representing physical domain (origin, spacing, direction, etc).
The second image has no metadata.
However, both have the same region, defined by an index (usually [0,0], but it can be different), and a size. So they both have the same amount of pixels.

With your patch, I cannot mix a "Point" from one image and the other one.
For example, a Point that was representing the index [2,0] (that can be used in **both** images, no matter the metadata), now it would be [-12, -10] if the Origin of the first image is `[-10, -10]`.

Before, that Point just represent a position in the Region (without physical space).

Also, in your approach, you are not using ITK `Spacing`, that is not always [1,1], as you assume in your patch. And also you would need to take into account the Direction, because 3D images can be tilted.

If you want to solve the problem of not having the image always centered in 0,0,0, we will need to introduce the notion of physical space versus index space, as ITK does.

**Comment by phcerdan** [comment]
If you want to read more about the physical domain, I recommend the [Software Guide of ITK, section 4.1.4](https://itk.org/ItkSoftwareGuide.pdf)
image

ITK has a set of function to convert between those: `TransformPhysicalPointToIndex()` and `TransformIndexToPhysicalPoint()` and I tell you after working for a while in ITK, the gold rule is that you have to work in only one of those spaces, Physical, or Index. You mix them, and bugs appear.

Your patch is mixing them.

**Comment by kerautret** [comment]
@phcerdan I know, actually the Domain in DGtal are able to handle floating coordinate and for this reason applying physical coordinates are not possible. The patch was just useful when the user want to get an isotropic volume with a given shit. If we want to fully handle ITK with spacing is not simple.

**Comment by kerautret** [comment]
but thank perhaps we can just add a flag to shift when the user want to shift the domain.

**Comment by phcerdan** [comment]
I think could be ok to just add a couple of functions to work in the Physical space, but I am afraid that the `operator()` is impossible to disambiguate between index space and physical space.

We can create a flag, and the `operator()` would behave differently depending on it, but it's going to create problems...
Better to have clear names to work in Physical space and leave `operator()` as it is (working in index space).

Could you point to where you are using this? Is it for visualization? Can you not use `operator()` but other function instead, like `GetPhysicalPoint()`? Or use operator() to get the index, and then apply the transformation to get the physical point and use that for visualization?

**Comment by kerautret** [comment]
> I think could be ok to just add a couple of functions to work in the Physical space, but I am afraid that the `operator()` is impossible to disambiguate between index space and physical space.

Agree, perhaps a solution could be to define some ImageAdapter to handle more specific ITK information like spacing etc. 

> We can create a flag, and the operator() would behave differently depending on it, but it's going to create problems...
Better to have clear names to work in Physical space and leave operator() as it is (working in index space).

Sure and actually no algorithms in DGtal are ready to handle for instance non regular spacing, but here it not physical coordinates but just the import of an image shift, like it already the case for other types of image well used in DGTal (.vol) that are not necessary centered in 0.

> Could you point to where you are using this? Is it for visualization? Can you not use operator() but other function instead, like GetPhysicalPoint()? Or use operator() to get the index, and then apply the transformation to get the physical point and use that for visualization?

I use it for instance when extracting digital surface from volumic file by using volBoundary2obj from DGtalTools and displaying it in itk-vtk-viewer that keep original physical coordinates of the source image. In online demonstration (like [here](
https://ipol-geometry.loria.fr/~kerautre/ipol_demo/LiverVesselnessIPOLDemo/result?key=70B6E0208EE8435F14B5802F5350B1E1)) it was also important to use through numerous tools without applying other changes after importing data.

**Comment by rolanddenis** [comment]
I also use non zero origin and non integer spacings with nD images and export them in VTK format. For now, I store the image and physical domain properties separately.

> Agree, perhaps a solution could be to define some ImageAdapter to handle more specific ITK information like spacing etc.

We could maybe define a physical space class that features (at least) a function to convert index to physical point. I'm not sure if the reverse way (physical to index) should be implemented and how: nearest point? neighbourhood (for interpolation purpose)?

> Sure and actually no algorithms in DGtal are ready to handle for instance non regular spacing, but here it not physical coordinates but just the import of an image shift, like it already the case for other types of image well used in DGTal (.vol) that are not necessary centered in 0.

The problem is that it has no meaning if spacing it not an integer (in my case, domain extent can also be less that 1). I would prefer that the user chooses to shift the image using a kind of `ImageAdapter` when he wants to, e.g.:
```C++
assert(std::abs(itk_image.getPhysicalSpacing() - 1.) <= 1e-12); // dummy check
auto shifted_itk_image = shiftImage(itk_image, itk_image.getPhysicalOrigin());
```

We could also wrap an image and a physical domain into a kind of physical image (not a `CImage` model) with a `operator() (RealPoint const&)` and a `operator[] (Point const&)` operators but it is probably not what we want here (not compatible with DGtal functions and how the `operator()` should be implemented, interpolation?).

**Comment by kerautret** [comment]
And what's about adding a new importWithDomainFunctor in the GenericReader ? By this way we can include something similar than the @rolanddenis (inside new functor) and we can remove my fix that will b fine for @phcerdan ? For now just a DomainShift function will be ok to to be added in the tools (but it will just add few lines to detect when ITK is used).

**Comment by phcerdan** [comment]
Please review the new commit. I have added a `shiftDomain` parameter. That can be provided only whith `updateDomain`, and defaults to `Zero`.

Current implementation is untested, and I would say it goes out of the buffer of ITK region when using  `operator()` or `setValue`.

**Comment by phcerdan** [comment]
This is probably not the best solution, but it is compatible (I think) with @kerautret needs.

I would be keen to review a better approach. But take this as a hot-patch.

**Comment by kerautret** [comment]
Many Thanks @phcerdan ! I look it!

**Comment by kerautret** [comment]
That looks nice, I tested your branch and all is perfectly working 👍🏻👏🏻
By testing I integrate it in the ITKReader that allows to have the same behavior as I use in the tools. But perhaps it is not perfect for you ? (I PR it in your branch ; https://github.com/phcerdan/DGtal/pull/1). 
Just Have. a look on the doc warning that I mention like travis.

**Comment by kerautret** [comment]
Thanks a lot for all @phcerdan for me it is fine just look the doc warning and we can merge. 👍🏻
I checked the doxygen generated page and perfect for me !

File: tests/images/testITKImage.cpp [hunk 1]
```diff
@@ -210,6 +210,184 @@ bool testITKMethod()
...
+  origin.Fill(10.0);
+  Image::ITKImage::SpacingType spacing;
+  spacing.Fill(2.0);
+  Image::ITKImage::DirectionType direction;
+  direction.SetIdentity();
+  itk_image->SetOrigin(origin);
+  itk_image->SetSpacing(spacing);
+  itk_image->SetDirection(direction);
+
+  // Check that getting index points is not affected by change of metadata.
```
Comment (kerautret) [inline tests/images/testITKImage.cpp:256]:
```suggestion
  // Check that getting value of index points is not affected by change of metadata.
```

File: src/DGtal/images/ImageContainerByITKImage.h [hunk 2]
```diff
@@ -305,6 +318,63 @@ namespace DGtal
...
 
+      // ----------------- DomainPoint to/from IndexPoint interface -------------
+
+      /**
+       * IndexPoint refers to a valid ITKImage::IndexPoint
+       * DomainPoint refers to a Point between lowerBound and upperBound of myDomain.
+       * They are different only when myDomainShift is different than Zero.
+       *
+       * @param indexPoint
+       *
```
[the remark this answers is not in the document]
Comment (phcerdan) [reply src/DGtal/images/ImageContainerByITKImage.h:329]: Added some description.
```
* @param indexPoint a point holding valid index coordinates of the ITK image

```

File: src/DGtal/images/ImageContainerByITKImage.h [hunk 3]
```diff
@@ -305,6 +318,63 @@ namespace DGtal
...
+      inline Point getDomainPointFromIndex(const Point &indexPoint) const;
+
+      inline Point getIndexFromDomainPoint(const Point &domainPoint) const;
+
+      /**
+       * The same as @getDomainPointFromItkIndex and @getIndexFromDomainPoint
+       * but using ITK types.
+       *
+       * @param itkIndexPoint
+       *
```
[the remark this answers is not in the document]
Comment (phcerdan) [reply src/DGtal/images/ImageContainerByITKImage.h:341]: Thanks, fixed: 
```
* @param itkIndexPoint an IndexType of ITK.
```

File: src/DGtal/images/ImageContainerByITKImage.h [hunk 4]
```diff
@@ -45,6 +45,7 @@
 #include "DGtal/base/CLabel.h"
```
[the remark this answers is not in the document]
Comment (phcerdan) [reply src/DGtal/images/ImageContainerByITKImage.h:45]: Done, thanks.

File: src/DGtal/images/ImageContainerByITKImage.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (phcerdan) [reply src/DGtal/images/ImageContainerByITKImage.h:329]: Nice, remove extra white lines

File: tests/images/testITKImage.cpp (hunk 1 above)
Comment (phcerdan) [reply tests/images/testITKImage.cpp:256]: Corrected, thanks.

File: src/DGtal/images/ImageContainerByITKImage.ih [hunk 5]
```diff
@@ -200,10 +199,124 @@ namespace DGtal
...
+    template<typename TDomain, typename TValue>
+    inline
+    typename ImageContainerByITKImage<TDomain, TValue>::Point
+    ImageContainerByITKImage<TDomain, TValue>::getDomainPointFromIndex ( const Point & indexPoint ) const
+    {
+      Point dgtal_domain_point = indexPoint;
+      for (Dimension k = 0; k < dimension; k++)
+        dgtal_domain_point[k] = indexPoint[k] + myDomainShift[k];
+
+      return dgtal_domain_point;
```
Comment (rolanddenis) [inline src/DGtal/images/ImageContainerByITKImage.ih:217]: Why not simply:
```C++
{
  return indexPoint + myDomainShift;
}
```

(same for other methods)

Comment (phcerdan) [reply src/DGtal/images/ImageContainerByITKImage.ih]: Nice catch, I can only do this in the functions that do not perform a change of type or mix ITK and DGtal types.

Applied to `getDomainFromIndex` and `getIndexFromDomain`.
Thanks

**Comment by kerautret** [comment]
@rolanddenis ok for merging ?

**Comment by rolanddenis** [comment]
Ok for me too! So `operator()` is same as before (I mean before the reversed PR) but there is now additional methods to convert from index to physical points and vice-versa, right?

**Comment by kerautret** [comment]
> Ok for me too! So operator() is same as before (I mean before the reversed PR) but there is now additional methods to convert from index to physical points and vice-versa, right?

Yes, and I have tested by integrated it in the Reader and by using a default parameter set to true I obtain the same behavior without changing the DGtalTools I will pu the PR between two meeting I hope today :)

**Comment by kerautret** [comment]
All is fine merging 🎉Thanks a lot @phcerdan

**Pull request merged**